pubsub/producer.go

```diff
@@ -216,15 +148,24 @@ func (p *Producer[Request, Response]) checkResponses(ctx context.Context) time.D
 			return 0
 		}
 		checked++
-		res, err := p.client.Get(ctx, id).Result()
+		msgKey := MessageKeyFor(p.redisStream, id)
+		res, err := p.client.Get(ctx, msgKey).Result()
 		if err != nil {
-			errSetId := setMinIdInt(&minIdInt, id)
-			if errSetId != nil {
-				log.Error("redis producer: error setting minId", "err", err)
-				return p.cfg.CheckResultInterval
-			}
 			if !errors.Is(err, redis.Nil) {
-				log.Error("redis producer: Error reading value in redis", "key", id, "error", err)
+				log.Error("Error reading value in redis", "key", msgKey, "error", err)
+			} else {
+				// The request this producer is waiting for has been past its TTL or is older than current PEL's lower,
+				// so safe to error and stop tracking this promise
+				allowedOldestID := fmt.Sprintf("%d-0", time.Now().Add(-p.cfg.RequestTimeout).UnixMilli())
+				if pelData != nil && pelData.Lower != "" {
+					allowedOldestID = pelData.Lower
+				}
+				if cmpMsgId(id, allowedOldestID) == -1 {
+					promise.ProduceError(errors.New("error getting response, request has been waiting for too long"))
+					log.Error("error getting response, request has been waiting past its TTL")
```

Comment:
I think theres a flaw here, instead setting allowedOldestId to lower (when this exists) we should only check for if a message has been waiting past its TTL irrespective of lower.

Because the second option (lower) should never hit unless a worker answered a message and acked it, but redis somehow lost the response or the producer for some reason couldn't read from redis (both shouldn't really happen)